client/asset/btc/btc.go

```diff
@@ -1136,132 +1153,320 @@ func (btc *ExchangeWallet) LocktimeExpired(contract dex.Bytes) (bool, time.Time,
+// looking for the maker's redemption beginning at swapconf confirmations
+// regardless of whether the server sends the 'redemption' message or not.
+func (btc *ExchangeWallet) FindRedemption(coinID dex.Bytes) (chan *asset.FindRedemptionResult, error) {
+	btc.findRedemptionMtx.Lock()
+	defer btc.findRedemptionMtx.Unlock()
+
 	txHash, vout, err := decodeCoinID(coinID)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("cannot decode contract coin id: %v", err)
 	}
-	txid := txHash.String()
-	// Get the wallet transaction.
-	tx, err := btc.wallet.GetTransaction(txHash.String())
+	contractOp := newOutPoint(txHash, vout)
+	if existingReq, inQueue := btc.findRedemptionQueue[contractOp]; inQueue {
+		btc.log.Debugf("duplicate FindRedemption request for %s", contractOp.String())
+		return existingReq.resultChan, nil
+	}
+	rawTx, err := btc.node.GetRawTransactionVerbose(txHash)
 	if err != nil {
-		if isTxNotFoundErr(err) {
-			return nil, asset.CoinNotFoundError
-		}
-		return nil, fmt.Errorf("error finding transaction %s in wallet: %v", txHash, err)
+		return nil, fmt.Errorf("error getting contract details: %v", err)
 	}
-	if tx.BlockIndex == 0 {
-		// This is a mempool transaction, GetRawTransactionVerbose works on mempool
-		// transactions, so grab the contract from there and prepare a hash.
-		rawTx, err := btc.node.GetRawTransactionVerbose(txHash)
-		if err != nil {
-			return nil, fmt.Errorf("error getting contract details from mempool")
-		}
-		if int(vout) > len(rawTx.Vout)-1 {
-			return nil, fmt.Errorf("vout index %d out of range for transaction %s", vout, txHash)
-		}
-		contractHash, err := dexbtc.ExtractContractHash(rawTx.Vout[vout].ScriptPubKey.Hex)
+	if int(vout) > len(rawTx.Vout)-1 {
+		return nil, fmt.Errorf("vout index %d out of range for transaction %s", vout, txHash)
+	}
+	contractHash, err := dexbtc.ExtractContractHash(rawTx.Vout[vout].ScriptPubKey.Hex)
+	if err != nil {
+		return nil, err
+	}
+	var contractBlock *block
+	if rawTx.BlockHash != "" {
+		blk, err := btc.getBlockHeader(rawTx.BlockHash)
 		if err != nil {
-			return nil, err
+			return nil, fmt.Errorf("getBlockHeader error for hash %q: %v", rawTx.BlockHash, err)
 		}
-		// Only need to look at other mempool transactions. For each txid, get the
-		// verbose transaction and check each input's previous outpoint.
-		txs, err := btc.node.GetRawMempool()
+		contractBlock = &block{hash: rawTx.BlockHash, height: blk.Height}
+	}
+
+	resultChan := make(chan *asset.FindRedemptionResult, 1)
+	btc.findRedemptionQueue[contractOp] = &findRedemptionReq{
+		contractHash: contractHash,
+		resultChan:   resultChan,
+	}
+
+	// Mempool contracts may only be spent by another mempool tx.
+	if contractBlock == nil {
+		go btc.findRedemptionInMempool(contractOp)
+		return resultChan, nil
+	}
+
+	// Begin searching for redemption for this contract from the block
+	// in which this contract was mined up till the current best block.
+	btc.tipMtx.RLock()
+	bestBlock := btc.currentTip
```

Comment:
This is a pointer so the value may change after copying the pointer. Should copy the value and use the new pointer.